### DISCUSSION TITLE
Dynamic next.config.js

### DISCUSSION THREAD

**Discussion by Amerehei** [vercel/next.js#60131] [App Router]
I use `process.env` to config `remotePatterns` in the `next.config.js`.
I would like to use my docker image in different environment. When I build the next.js app, next.js embed the value of `process.env` to my standalone server.js file at the build time. How can we change the behaviour and keep the ENV for in the runtime?

How can we dynamically change `next.config.js` for different env?

**Comment by NickyBall** [comment]
I think maybe you can mount the next.config.js file from the host to container.
![image](https://github.com/vercel/next.js/assets/5468446/d02ad502-4b5c-41b0-ad2b-23af3b7694fa)

**Reply by Amerehei** [reply]
Are you sure? Do we have next.config.js on runtime?

**Comment by KoenBrouwer** [comment]
I'm facing the same issue, but I'd like to add that this is only a problem when we set `output: "standalone"` in next.config.js. That's when `next.config.js` gets evaluated and embedded in `standalone/server.js`. 
Would this maybe mean that we can't use standalone output if we need this kind of runtime configuration?

**Comment by axelrindle** [comment]
I'm currently messing with post-processing the generated `server.js` file where I replace the `const nextConfig = {}` line with an import statement to another js file. I will post this here if I get this right.

But it is stuff like this which always makes me regret choosing Next.js in a project. Coming from Laravel where everything can be configured dynamically at runtime, Next.js is such a pain to work with.

**Reply by axelrindle** [reply]
Ok I basically got it working now.

https://gist.github.com/axelrindle/1d76618caf84d62caba429adf5381f56

After `next build` run the `modify-config.ts` script (I use Bun: `bun run modify-config.ts`). It will patch the standalone `server.js` to call the custom `makeConfig` function which patches the embedded Next.js config object.

**Reply by jkowalc** [reply]
`server.js` is not the only place where the config is placed.

From what I gathered so far:
- A copy of the config is placed under `.next/required-server-files.json` (not sure if not updating this breaks things)
- Values of certain config options may be placed all over the server files. From my experiments this applies to at least `basePath` and `assetPrefix`, but there could be more. There is unfortunately no documentation as to which options behave like this.

Overall this doesn't look like a stable solution. Have you run this in production?
